lib/handler/mruby/sleep.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2017 Ichito Nagata
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include <mruby.h>
+#include <mruby/array.h>
+#include <mruby/error.h>
+#include "h2o/mruby_.h"
+
+struct st_h2o_mruby_sleep_context_t {
+    h2o_mruby_context_t *ctx;
+    mrb_value receiver;
+    h2o_timeout_t timeout;
+    h2o_timeout_entry_t timeout_entry;
+    uint64_t started_at;
+};
+
+static void on_deferred_timeout(h2o_timeout_entry_t *entry)
+{
+    struct st_h2o_mruby_sleep_context_t *ctx = H2O_STRUCT_FROM_MEMBER(struct st_h2o_mruby_sleep_context_t, timeout_entry, entry);
+    h2o_timeout_dispose(ctx->ctx->shared->ctx->loop, &ctx->timeout);
+    free(ctx);
+}
+
+static void on_sleep_timeout(h2o_timeout_entry_t *entry)
+{
+    struct st_h2o_mruby_sleep_context_t *ctx = H2O_STRUCT_FROM_MEMBER(struct st_h2o_mruby_sleep_context_t, timeout_entry, entry);
+    assert(! mrb_nil_p(ctx->receiver));
+    h2o_mruby_shared_context_t *shared = ctx->ctx->shared;
+    mrb_int sleep_sec = (mrb_int)(h2o_now(shared->ctx->loop) - ctx->started_at) / 1000;
+
+    h2o_mruby_run_fiber(ctx->ctx, ctx->receiver, mrb_fixnum_value(sleep_sec), NULL);
+
+    mrb_gc_unregister(shared->mrb, ctx->receiver);
+    h2o_timeout_unlink(entry);
+
+    /* defer freeing to avoid concurrent modification onf timeout linklist */
+    entry->cb = on_deferred_timeout;
+    h2o_timeout_link(shared->ctx->loop, &shared->ctx->zero_timeout, entry);
```

Comment:
I think that we should resolve the issue after merging the PR using the current approach.

Basically what you need to do is record the address of `h2o_timeout_t` that is being processed by `h2o_evloop_run` as a member of `h2o_evloop_t`, and if the address matches that of the disposed timeout, then update the recorded address so that it would point to the previous object. However implementing this correctly might require some work.